### PR TITLE
Feature: Add global UDFs to repository

### PR DESCRIPTION
**Pull request by wyatt-joyner-tg**
These global UDFs will cover all UDF dependencies for all algorithms provided in this repository.

### CODE REVIEW

File: UDF/tg_ExprFunctions.hpp [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+#include <random>
+#include <vector>
+
+/**     XXX Warning!! Put self-defined struct in ExprUtil.hpp **
+ *  No user defined struct, helper functions (that will not be directly called
+ *  in the GQuery scripts) etc. are allowed in this file. This file only
+ *  contains user-defined expression function's signature and body.
+ *  Please put user defined structs, helper functions etc. in ExprUtil.hpp
+ */
+#include "ExprUtil.hpp"
```
Comment (romansod) [inline UDF/tg_ExprFunctions.hpp:55]: Should be "tg_ExprUtil.hpp"

File: UDF/tg_ExprFunctions.hpp [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+
+/**     XXX Warning!! Put self-defined struct in ExprUtil.hpp **
+ *  No user defined struct, helper functions (that will not be directly called
+ *  in the GQuery scripts) etc. are allowed in this file. This file only
+ *  contains user-defined expression function's signature and body.
+ *  Please put user defined structs, helper functions etc. in ExprUtil.hpp
+ */
+#include "tg_ExprUtil.hpp"
+
+namespace tg_UDIMPL {
```
Comment (romansod) [inline UDF/tg_ExprFunctions.hpp:57]: This needs to be just UDIMPL. Right now we only support the single namespace. There are checks in my transaction during installation that make sure there is no ambiguity

File: UDF/tg_ExprFunctions.hpp [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+ *
+ * - Please don't remove necessary codes in this file
+ *
+ * - A backup of this file can be retrieved at
+ *     <tigergraph_root_path>/dev_<backup_time>/gdk/gsql/src/QueryUdf/ExprFunctions.hpp
+ *   after upgrading the system.
+ *
+ ******************************************************************************/
+
+#ifndef EXPRFUNCTIONS_HPP_
```
Comment (romansod) [inline UDF/tg_ExprFunctions.hpp:32]: This should probably have a different name from the regular ExprFunctions.hpp file. Maybe TG_EXPRFUNTIONS_HPP_

File: UDF/tg_ExprFunctions.hpp [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+ *  Please put user defined structs, helper functions etc. in ExprUtil.hpp
+ */
+#include "tg_ExprUtil.hpp"
+
+namespace tg_UDIMPL {
+  typedef std::string string; //XXX DON'T REMOVE
+
+  /****** BIULT-IN FUNCTIONS **************/
+  /****** XXX DON'T REMOVE ****************/
+  inline int str_to_int (string str) {
```
Comment (romansod) [inline UDF/tg_ExprFunctions.hpp:62]: I don't think we want to include these. These should be in ExprFunctions already and will likely result in conflict. I only see std::to_string used in this file, so removing them should not be problematic. Compile it to make sure once you remove them

File: UDF/tg_ExprFunctions.hpp [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+
+  inline string to_string (double val) {
+    char result[200];
+    sprintf(result, "%g", val);
+    return string(result);
+  }
+
+  /* ============== START SPEAKER LISTENER LABEL PROP ================== */
+
+  inline double getVal (ListAccum<double>& list, int64_t i) {
```
Comment (romansod) [inline UDF/tg_ExprFunctions.hpp:78]: Any function in this file can be referenced through GSQL, so all of these function names need the prefix. If there are helper functions here that we do not want user to call from GSQL, they should go in tg_ExprUtil.hpp

Comment (wyatt-joyner-tg) [reply UDF/tg_ExprFunctions.hpp:62]: Understood. Applying now.

File: UDF/tg_ExprUtil.hpp [hunk 6]
```diff
@@ -0,0 +1,718 @@
...
+ *
+ * - Please don't remove necessary codes in this file
+ *
+ * - A backup of this file can be retrieved at
+ *     <tigergraph_root_path>/dev_<backup_time>/gdk/gsql/src/QueryUdf/ExprUtil.hpp
+ *   after upgrading the system.
+ *
+ ******************************************************************************/
+
+#ifndef EXPRUTIL_HPP_
```
Comment (romansod) [inline UDF/tg_ExprUtil.hpp:20]: This should also be TG_EXPRUTIL_HPP_

File: UDF/tg_ExprUtil.hpp [hunk 7]
```diff
@@ -0,0 +1,718 @@
...
+ * - This library is for defining struct and helper functions that will be used
+ *   in the user-defined functions in "ExprFunctions.hpp". Note that functions
+ *   defined in this file cannot be directly called from TigerGraph Query scripts.
+ *   Please put such functions into "ExprFunctions.hpp" under the same directory
+ *   where this file is located.
+ *
+ * - Please don't remove necessary codes in this file
+ *
+ * - A backup of this file can be retrieved at
+ *     <tigergraph_root_path>/dev_<backup_time>/gdk/gsql/src/QueryUdf/ExprUtil.hpp
```
Comment (romansod) [inline UDF/tg_ExprUtil.hpp:15]: Path is good, just change the end to tg_ExprUtil.hpp

File: UDF/tg_ExprUtil.hpp [hunk 8]
```diff
@@ -0,0 +1,718 @@
+/******************************************************************************
+ * Copyright (c) 2021, TigerGraph Inc.
+ * All rights reserved.
+ * Project: TigerGraph Query Language
+ *
+ * - This library is for defining struct and helper functions that will be used
+ *   in the user-defined functions in "ExprFunctions.hpp". Note that functions
+ *   defined in this file cannot be directly called from TigerGraph Query scripts.
+ *   Please put such functions into "ExprFunctions.hpp" under the same directory
```
Comment (romansod) [inline UDF/tg_ExprUtil.hpp:9]: "tg_ExprFunctions.hpp"

File: UDF/tg_ExprFunctions.hpp [hunk 9]
```diff
@@ -0,0 +1,420 @@
...
+ *     - string (don't use std::string)
+ *     - accumulators
+ *
+ * - Function names are case sensitive, unique, and can't be conflict with
+ *   built-in math functions and reserve keywords.
+ *
+ * - Please don't remove necessary codes in this file
+ *
+ * - A backup of this file can be retrieved at
+ *     <tigergraph_root_path>/dev_<backup_time>/gdk/gsql/src/QueryUdf/ExprFunctions.hpp
```
Comment (romansod) [inline UDF/tg_ExprFunctions.hpp:27]: Path is good, just update to tg_ExprFunctions.hpp

File: UDF/tg_ExprFunctions.hpp [hunk 10]
```diff
@@ -0,0 +1,420 @@
+/******************************************************************************
+ * Copyright (c) 2015-2021, TigerGraph Inc.
+ * All rights reserved.
+ * Project: TigerGraph Query Language
+ * udf.hpp: a library of user defined functions used in queries.
+ *
+ * - This library should only define functions that will be used in
+ *   TigerGraph Query scripts. Other logics, such as structs and helper
+ *   functions that will not be directly called in the GQuery scripts,
+ *   must be put into "ExprUtil.hpp" under the same directory where
```
Comment (romansod) [inline UDF/tg_ExprFunctions.hpp:10]: "tg_ExprUtil.hpp"

File: UDF/tg_ExprFunctions.hpp [hunk 11]
```diff
@@ -0,0 +1,420 @@
...
+    return similarity;
+  }
+
+  /* =========== END APPROXIMATE NEAREST NEIGHBORS ============= */
+
+}
+
+/****************************************/
+
+#endif /* EXPRFUNCTIONS_HPP_ */
```
Comment (romansod) [inline UDF/tg_ExprFunctions.hpp:420]: TG_EXPRFUNCTIONS_HPP_

File: UDF/tg_ExprUtil.hpp [hunk 12]
```diff
@@ -0,0 +1,718 @@
...
+ *     int age;
+ *     double height;
+ *     double weight;
+ *   }
+ *
+ */
+}
+
+
+#endif /* EXPRUTIL_HPP_ */
```
Comment (romansod) [inline UDF/tg_ExprUtil.hpp:718]: TG_EXPRUTIL_HPP_

File: UDF/tg_ExprFunctions.hpp [hunk 13]
```diff
@@ -0,0 +1,420 @@
...
+
+
+  /* ============== START FREQUENT PATTERN MINING ===================== */
+
+  inline int64_t tg_get_hash(int64_t id){
+	    int64_t hash_val = id % tg::MOD;
+	    return hash_val;
+	}
+
+	template<typename tup>
```
Comment (romansod) [inline UDF/tg_ExprFunctions.hpp:90]: Indentation

File: UDF/tg_ExprUtil.hpp [hunk 14]
```diff
@@ -0,0 +1,717 @@
...
+  /* ======= END FREQUENT PATTERN MINING ======== */
+
+
+
+  /* ======== START NODE2VEC ==============*/
+  typedef std::vector<float> Vector;
+
+  struct Word
+  {
+          int32_t index_;
```
Comment (romansod) [inline UDF/tg_ExprUtil.hpp:62]: Indentation

**Comment by romansod** [comment]
Once you make the indentation consistent I will approve

**Comment by wyatt-joyner-tg** [comment]
> Once you make the indentation consistent I will approve

Done!

**Comment by wyatt-joyner-tg** [comment]
Conflicts resolved.

**Review by songrenchu — APPROVED** [review]
Approve for the files other than tg_ExprFunctions.hpp and tg_ExprUtil.hpp

**Pull request merged**